src/communication/v2/session.hpp

```diff
@@ -0,0 +1,347 @@
+#include "communication/exceptions.hpp"
+#include "utils/logging.hpp"
+#include "utils/message.hpp"
+#include "utils/variant_helpers.hpp"
+
+namespace memgraph::communication::v2 {
+
+/**
+ * This is used to provide input to user Sessions. All Sessions used with the
+ * network stack should use this class as their input stream.
+ */
+using InputStream = communication::Buffer::ReadEnd;
+using tcp = boost::asio::ip::tcp;
+
+/**
+ * This is used to provide output from user Sessions. All Sessions used with the
+ * network stack should use this class for their output stream.
+ */
+class OutputStream final {
+ public:
+  explicit OutputStream(std::function<bool(const uint8_t *, size_t, bool)> write_function)
+      : write_function_(write_function) {}
+
+  OutputStream(const OutputStream &) = delete;
+  OutputStream(OutputStream &&) = delete;
+  OutputStream &operator=(const OutputStream &) = delete;
+  OutputStream &operator=(OutputStream &&) = delete;
+  ~OutputStream() = default;
+
+  bool Write(const uint8_t *data, size_t len, bool have_more = false) { return write_function_(data, len, have_more); }
+
+  bool Write(const std::string &str, bool have_more = false) {
+    return Write(reinterpret_cast<const uint8_t *>(str.data()), str.size(), have_more);
+  }
+
+ private:
+  std::function<bool(const uint8_t *, size_t, bool)> write_function_;
+};
+
+/**
+ * This class is used internally in the communication stack to handle all user
+ * Sessions. It handles socket ownership, inactivity timeout and protocol
+ * wrapping.
+ */
+template <typename TSession, typename TSessionData>
+class Session final : public std::enable_shared_from_this<Session<TSession, TSessionData>> {
+  using TCPSocket = boost::asio::ip::tcp::socket;
+  using SSLSocket = boost::asio::ssl::stream<TCPSocket>;
+  using std::enable_shared_from_this<Session<TSession, TSessionData>>::shared_from_this;
+
+ public:
+  template <typename... Args>
+  static std::shared_ptr<Session> Create(Args &&...args) {
+    return std::shared_ptr<Session>(new Session(std::forward<Args>(args)...));
+  }
+
+  Session(const Session &) = delete;
+  Session(Session &&) = delete;
+  Session &operator=(const Session &) = delete;
+  Session &operator=(Session &&) = delete;
+  ~Session() {
+    if (IsConnected()) {
+      spdlog::error("Session: Destructor called while execution is active");
+    }
+  }
+
+  bool Start() {
+    if (IsConnected()) {
+      return false;
+    }
+    execution_active_ = true;
+    timeout_timer_.async_wait(boost::asio::bind_executor(strand_, std::bind(&Session::OnTimeout, shared_from_this())));
+
+    if (auto *socket = std::get_if<SSLSocket>(&socket_); socket) {
+      boost::asio::dispatch(shared_from_this()->strand_,
+                            [shared_this = shared_from_this()] { shared_this->DoHandshake(); });
+    } else {
+      boost::asio::dispatch(this->shared_from_this()->strand_,
+                            [shared_this = shared_from_this()] { shared_this->DoRead(); });
```

Comment:
```suggestion
      boost::asio::dispatch(strand_,
                            [shared_this = shared_from_this()] { shared_this->DoHandshake(); });
    } else {
      boost::asio::dispatch(strand_,
                            [shared_this = shared_from_this()] { shared_this->DoRead(); });
```

We don't need to create a `shared_ptr` just to get the strand.